### PR TITLE
White Box Attackers

### PR DESCRIPTION
**Pull request by divyahansg**
- experiment evaluation files `(exp_)`
- HillClimbingAttacker, LowDegreeAttacker, HighDegreeAttacker 
- caches mapping of item -> degree in `.npy`

### CODE REVIEW

File: exp_final_whitebox_attacker.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+from gbra.data.network_loader import Movielens100kLoader
+from gbra.attackers.attacker import *
+from gbra.recommender.recommenders import PixieRandomWalkRecommender
+
+ITERATIONS = 10
+
+PIXIE_PARAMS = {
+    'n_p': 30,
+    'n_v': 4,
+    'num_steps_in_walk': 200,
```
Comment (alanefl) [inline exp_final_whitebox_attacker.py]: 1000?

File: gbra/util/ei_graph.py [hunk 2]
```diff
@@ -234,7 +240,7 @@ def load(filename):
             possible_ratings = sorted(list(ratings_set))
 
         # Setup the graph with the range of possible ratings.
-        graph.rating_range = (possible_ratings[0], possible_ratings[1])
+        graph.rating_range = (0, 5)
```
Comment (alanefl) [inline gbra/util/ei_graph.py]: The rating range for movielens is actually (1, 5). 

The line above actually has a bug.  It should be `graph.rating_range = (possible_ratings[0], possible_ratings[-1])`.

File: gbra/recommender/recommenders.py [hunk 3]
```diff
@@ -244,8 +244,13 @@ def _sample_walk_length(self):
         Intuitively, smaller alphas bias towards shorter walks, whereas
         larger alphas bias towards longer walks.
         """
+<<<<<<< HEAD
+        mu = int(round(self._alpha * self._num_steps_in_walk))
```
Comment (alanefl) [inline gbra/recommender/recommenders.py]: Looks like there's some merge conflicts.  The issues here are 1) I renamed `num_steps_in_walk` to `max_steps_in_walk` and 2) I added a new parameter `beta` to the recommender which is effectively the value of `sigma`.

Comment (divyahansg) [reply gbra/recommender/recommenders.py]: oh i thought i resolved the merge

File: exp_final_whitebox_attacker.py (hunk 1 above)
Comment (divyahansg) [reply exp_final_whitebox_attacker.py]: fixed

File: exp_final_whitebox_attacker.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+attackers = {
+    'HighDegreeAttacker': HighDegreeAttacker,
+    'LowDegreeAttacker': LowDegreeAttacker,
+    'HillClimbingAttacker': HillClimbingAttacker
+}
+
+def get_attacker(network, recommender, target_item):
+    attacker_klass = attackers[ATTACKER_NAME]
+    num_fake_entities = int(percent_fake_entities * network.num_entities)
```
Comment (alanefl) [inline exp_final_whitebox_attacker.py]: `percent_fake_entities` -> `PERCENT_FAKE_ENTITIES`?

File: exp_final_whitebox_attacker.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    return attacker_klass(recommender, target_item, num_fake_entities, NUM_FAKE_REVIEWS)
+
+def evaluate_attacker():
+    network = Movielens100kLoader().load()
+
+    # Target item to popularize
+    [target_item] = network.get_random_items(1)
```
Comment (alanefl) [inline exp_final_whitebox_attacker.py]: Instead of getting a random item could we choose a fixed and relatively unpopular item?  We should look at the same item across all attackers and experiments.

File: exp_final_whitebox_attacker.py (hunk 4 above)
Comment (divyahansg) [reply exp_final_whitebox_attacker.py]: fixed

File: exp_final_whitebox_attacker.py (hunk 1 above)
Comment (alanefl) [reply exp_final_whitebox_attacker.py]: Also with the new merged code this is not `num_steps_in_walk` but `max_steps_in_walk`

File: exp_final_whitebox_attacker.py (hunk 5 above)
Comment (alanefl) [reply exp_final_whitebox_attacker.py]: (this is OK because we are sampling across 10 iterations)

File: exp_final_whitebox_attacker.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+target_items = network.get_random_items(ITERATIONS)
+print target_items
+for i in range(ITERATIONS):
+    (before, after) = evaluate_attacker(target_items[i])
+    print (before, after)
+    results.append((before, after))
+print results
+
+arr = np.array([a[1] for a in results])
+print "mean: %f, median %f" % (np.mean(arr), np.median(arr))
```
Comment (FluxLemur) [inline exp_final_whitebox_attacker.py]: would have been good to see standard deviation -- might illustrate clearly whether we are indeed sampling enough ITERATIONS

File: gbra/util/ei_graph.py [hunk 7]
```diff
@@ -234,7 +240,7 @@ def load(filename):
             possible_ratings = sorted(list(ratings_set))
 
         # Setup the graph with the range of possible ratings.
-        graph.rating_range = (possible_ratings[0], possible_ratings[1])
+        graph.rating_range = (1, 5)
```
Comment (FluxLemur) [inline gbra/util/ei_graph.py]: why did you change this line, looked correct before?

Comment (divyahansg) [reply gbra/util/ei_graph.py]: https://github.com/alanefl/graph-based-recommender-attacks/pull/17#pullrequestreview-182894696

File: exp_final_whitebox_attacker.py (hunk 6 above)
Comment (divyahansg) [reply exp_final_whitebox_attacker.py]: a big std dev would indicate that some nodes are more susceptible than others, not really related to sampling; we'll have enough output to calculate that nayways
